File: web/src/core/shadowing.ts
import { PracticeEngine, prioritise } from './practice'
import type { CEFRLevel, CoursePack, Exercise, Lesson, UserState } from './types'

/**
 * Shadowing: hear the phrase, say it out loud, record yourself, hear both back.
 * Kept in step with `EnglishCoachCore/Shadowing.swift`.
 *
 * Everything else in the app is read with the eyes and answered with the fingers,
 * which never gets the mouth moving. There is no speech grading here and no need
 * for one: the learner hears their own take next to the model phrase and judges it.
 * The verdict is theirs, but it still feeds points and spaced repetition, so a
 * phrase that did not come out comes back.
 *
 * The phrases are the ones that already ship — no new content, no network, no keys.
 */

export const SHADOWING_LESSON_ID = 'shadowing'
const DEFAULT_SIZE = 8

export interface ShadowingItem {
  /** The exercise the phrase came from: progress is recorded against it. */
  exerciseID: string
  /** The English sentence to say out loud. */
  text: string
  /** Russian meaning, when the source exercise carries one. */
  gloss?: string
}

export interface ShadowingSet {
  items: ShadowingItem[]
  /** The same phrases as exercises, so the session records against real ids. */
  exercises: Exercise[]
}

export interface ShadowingOptions {
  courses: CoursePack[]
  level: CEFRLevel
  state: UserState
  size?: number
  now?: Date
  random?: () => number
}

const CYRILLIC = /[Ѐ-ӿ]/

/**
 * The line has to survive being read out loud. Content sometimes glosses a word in
 * Russian inside the sentence ("That is my bag (там, далеко)."): that is for the eye.
 * Anything still Russian after the glosses are dropped is not a phrase to shadow.
 */
function sayable(text: string): string | null {
  const spoken = text.replace(/\s*\([^)]*[Ѐ-ӿ][^)]*\)/g, '').replace(/\s+/g, ' ').trim()
  return spoken && !CYRILLIC.test(spoken) ? spoken : null
}

/** "We ___ this film before." + "have seen" → a whole sentence worth saying. */
function filledGap(exercise: Exercise): string | undefined {
  const prompt = exercise.prompt ?? ''
  const option = exercise.correctOption ?? ''
  if (!option || !/_{2,}/.test(prompt)) return undefined
  const sentence = prompt.replace(/_{2,}/, option)
  // A second gap would leave the sentence unsayable, so it is left out entirely.
  return /_{2,}/.test(sentence) ? undefined : sentence
}

/** The English sentence hidden in an exercise, or null when it has none to say. */
export function shadowingPhrase(exercise: Exercise): ShadowingItem | null {
  const make = (text: string | undefined, gloss?: string): ShadowingItem | null => {
    const spoken = sayable(text ?? '')
    if (!spoken) return null
    const meaning = (gloss ?? '').trim()
    return { exerciseID: exercise.id, text: spoken, gloss: meaning || undefined }
  }
  switch (exercise.type) {
    case 'flashcard':
      return make(exercise.prompt, exercise.translation)
    case 'translate':
    case 'word_order':
      return make(exercise.canonicalAnswer, exercise.prompt)
    case 'multiple_choice':
      return make(filledGap(exercise))
    default:
      return null // a rule card has nothing to say out loud
  }
}

export const ShadowingEngine = {
  /** Every phrase the learner could shadow at this level and below. */
  pool(courses: CoursePack[], level: CEFRLevel): Exercise[] {
    return PracticeEngine.pool(courses, level).filter((exercise) => shadowingPhrase(exercise) !== null)
  },

  count(courses: CoursePack[], level: CEFRLevel): number {
    return this.pool(courses, level).length
  },

  build({ courses, level, state, size = DEFAULT_SIZE, now = new Date(), random = Math.random }: ShadowingOptions): ShadowingSet {
    const exercises = prioritise(this.pool(courses, level), state, now, random).slice(0, size)
    return { exercises, items: exercises.map((exercise) => shadowingPhrase(exercise)!) }
  },

  /** Speaking is slower than typing, so a set is short; it is never "completed". */
  lesson(exercises: Exercise[]): Lesson {
    return {
      id: SHADOWING_LESSON_ID,
      title: 'Вслух за диктором',
      summary: 'Слушай, повторяй, сравнивай.',
      estimatedMinutes: Math.max(2, exercises.length),
      exercises,
    }
  },
}

File: native/Sources/EnglishCoachCore/Shadowing.swift
import Foundation

/// Shadowing: hear the phrase, say it out loud, record yourself, hear both back.
/// Kept in step with `web/src/core/shadowing.ts`.
///
/// Everything else in the app is read with the eyes and answered with the fingers,
/// which never gets the mouth moving. There is no speech grading here and no need
/// for one: the learner hears their own take next to the model phrase and judges it.
/// The verdict is theirs, but it still feeds points and spaced repetition, so a
/// phrase that did not come out comes back.
///
/// The phrases are the ones that already ship — no new content, no network, no keys.
public struct ShadowingItem: Identifiable, Equatable, Sendable {
    /// The exercise the phrase came from: progress is recorded against it.
    public let exerciseID: String
    /// The English sentence to say out loud.
    public let text: String
    /// Russian meaning, when the source exercise carries one.
    public let gloss: String?

    public var id: String { exerciseID }

    public init(exerciseID: String, text: String, gloss: String?) {
        self.exerciseID = exerciseID; self.text = text; self.gloss = gloss
    }
}

public struct ShadowingSet: Equatable, Sendable {
    public let items: [ShadowingItem]
    /// The same phrases as exercises, so the session records against real ids.
    public let exercises: [Exercise]
}

public enum ShadowingEngine {
    public static let lessonID = "shadowing"
    public static let defaultSize = 8

    /// The English sentence hidden in an exercise, or nil when it has none to say.
    public static func phrase(for exercise: Exercise) -> ShadowingItem? {
        func make(_ text: String?, gloss: String? = nil) -> ShadowingItem? {
            guard let spoken = sayable(text ?? "") else { return nil }
            let meaning = (gloss ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            return ShadowingItem(exerciseID: exercise.id, text: spoken, gloss: meaning.isEmpty ? nil : meaning)
        }
        switch exercise.type {
        case .flashcard: return make(exercise.prompt, gloss: exercise.translation)
        case .translate, .wordOrder: return make(exercise.canonicalAnswer, gloss: exercise.prompt)
        case .multipleChoice: return make(filledGap(exercise))
        case .info: return nil // a rule card has nothing to say out loud
        }
    }

    private static let cyrillic = "[\u{0400}-\u{04FF}]"

    /// The line has to survive being read out loud. Content sometimes glosses a word in
    /// Russian inside the sentence ("That is my bag (там, далеко)."): that is for the eye.
    /// Anything still Russian after the glosses are dropped is not a phrase to shadow.
    private static func sayable(_ text: String) -> String? {
        let spoken = text
            .replacingOccurrences(of: "\\s*\\([^)]*\(cyrillic)[^)]*\\)", with: "", options: .regularExpression)
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !spoken.isEmpty, spoken.range(of: cyrillic, options: .regularExpression) == nil else { return nil }
        return spoken
    }

    /// "We ___ this film before." + "have seen" → a whole sentence worth saying.
    private static func filledGap(_ exercise: Exercise) -> String? {
        let prompt = exercise.prompt ?? ""
        let option = exercise.correctOption ?? ""
        guard !option.isEmpty, let gap = prompt.range(of: "_{2,}", options: .regularExpression) else { return nil }
        let sentence = prompt.replacingCharacters(in: gap, with: option)
        // A second gap would leave the sentence unsayable, so it is left out entirely.
        return sentence.range(of: "_{2,}", options: .regularExpression) == nil ? sentence : nil
    }

    /// Every phrase the learner could shadow at this level and below.
    public static func pool(courses: [CoursePack], level: CEFRLevel) -> [Exercise] {
        PracticeEngine.pool(courses: courses, level: level).filter { phrase(for: $0) != nil }
    }

    public static func count(courses: [CoursePack], level: CEFRLevel) -> Int {
        pool(courses: courses, level: level).count
    }

    public static func build(
        courses: [CoursePack],
        level: CEFRLevel,
        state: UserState,
        size: Int = defaultSize,
        now: Date = .now,
        shuffle: ([Exercise]) -> [Exercise] = { $0.shuffled() }
    ) -> ShadowingSet {
        let ordered = PracticeEngine.prioritise(pool(courses: courses, level: level), state: state, now: now, shuffle: shuffle)
        let exercises = Array(ordered.prefix(size))
        return ShadowingSet(items: exercises.compactMap { phrase(for: $0) }, exercises: exercises)
    }

    /// Speaking is slower than typing, so a set is short; it is never "completed".
    public static func lesson(_ exercises: [Exercise]) -> Lesson {
        Lesson(
            id: lessonID,
            title: "Вслух за диктором",
            summary: "Слушай, повторяй, сравнивай.",
            estimatedMinutes: max(2, exercises.count),
            exercises: exercises
        )
    }
}
